source/common/router/config_impl.cc

```diff
@@ -87,10 +88,34 @@ class HeaderHashMethod : public HashPolicyImpl::HashMethod {
   const Http::LowerCaseString header_name_;
 };
 
+class CookieHashMethod : public HashPolicyImpl::HashMethod {
+public:
+  CookieHashMethod(const std::string& key, long ttl_seconds) : key_(key), ttl_(ttl_seconds) {}
+
+  Optional<uint64_t> evaluate(const std::string&, const Http::HeaderMap& headers,
+                              const HashPolicy::AddCookieCallback add_cookie) const override {
+    Optional<uint64_t> hash;
+    std::string value = Http::Utility::parseCookieValue(headers, key_);
+
+    if (value.empty() && ttl_ != 0) {
+      value = add_cookie(key_, ttl_);
+      hash.value(HashUtil::xxHash64(value));
+
+    } else if (!value.empty()) {
+      hash.value(HashUtil::xxHash64(value));
+    }
+    return hash;
```

Comment:
So if there's no existing cookie and the TTL is 0, shouldn't we still hash, and just not set the cookie?

Reply:
Not here, because there's nothing to hash on. The Router::Filter does that already when none of the policies return a hash.